numba_cuda/numba/cuda/models.py

```diff
@@ -42,3 +44,48 @@ def __init__(self, dmm, fe_type):
 
 
 register_model(CUDADispatcher)(models.OpaqueModel)
+
+
+def _as_bfloat(value):
+    # Step 1: Reinterpret the input as u32 bits
+    u = struct.unpack("I", struct.pack("f", value))[0]
+
+    # Step 2: Truncate (or round, we choose truncate) last 16 bits
+    trunc = u >> 16
+
+    # Step 3: Unpack them back to Python floats
+    f = struct.unpack("f", struct.pack("I", trunc))[0]
+
+    return f
+
+
+class BfloatType(ir.types._BaseFloatType):
+    """Brain-float type"""
+
+    null = "0.0"
+    intrinsic_name = "bfloat"
+
+    def __str__(self):
+        return "bfloat"
+
+    def format_constant(self, value):
+        return ir.types._format_double(_as_bfloat(value))
+
+
+BfloatType._create_instance()
+
+
+@register_model(Bfloat16)
+class _model___nv_bfloat16(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        from numba.cuda.api import get_current_device
+
+        major, minor = get_current_device().compute_capability
+
+        # Blackwell device leverage latest nvvm (llvm 20+ dialect) which has
+        # bfloat type
+        if major >= 10:
+            be_type = BfloatType()
```

Comment:
Did you find out if this mirrors what NVCC / NVRTC do on Blackwell?

Reply:
[edea3c3](https://github.com/NVIDIA/numba-cuda/pull/376/commits/edea3c3e7022e8f6b32a748b3638801b8feb3380)